include/drivers/console.h

```diff
@@ -7,14 +7,69 @@
 #ifndef __CONSOLE_H__
 #define __CONSOLE_H__
 
-#include <stdint.h>
+#include <utils_def.h>
 
-int console_init(uintptr_t base_addr,
-		unsigned int uart_clk, unsigned int baud_rate);
-void console_uninit(void);
+#define CONSOLE_T_NEXT			(U(0) * REGSZ)
+#define CONSOLE_T_FLAGS			(U(1) * REGSZ)
+#define CONSOLE_T_PUTC			(U(2) * REGSZ)
+#define CONSOLE_T_GETC			(U(3) * REGSZ)
+#define CONSOLE_T_FLUSH			(U(4) * REGSZ)
+#define CONSOLE_T_DRVDATA		(U(5) * REGSZ)
+
+#define CONSOLE_FLAG_BOOT		BIT(0)
+#define CONSOLE_FLAG_RUNTIME		BIT(1)
+#define CONSOLE_FLAG_CRASH		BIT(2)
+/* Bits 3 to 7 reserved for additional scopes in future expansion. */
+#define CONSOLE_FLAG_SCOPE_MASK		((U(1) << 8) - 1)
+/* Bits 8 to 31 reserved for non-scope use in future expansion. */
+
+/* Returned by getc callbacks when receive FIFO is empty. */
+#define ERROR_NO_PENDING_CHAR		(-1)
+/* Returned by console_xxx() if no registered console implements xxx. */
+#define ERROR_NO_VALID_CONSOLE		(-128)
+
+#ifndef __ASSEMBLY__
+
+#include <types.h>
+
+typedef struct console {
+	struct console *next;
+	u_register_t flags;
+	int (*putc)(int character, struct console *console);
+	int (*getc)(struct console *console);
+	int (*flush)(struct console *console);
+	/* Additional private driver data may follow here. */
+} console_t;
+#include <console_assertions.h>	/* offset macro assertions for console_t */
+
+/*
+ * NOTE: There is no publicly accessible console_register() function. Consoles
+ * are registered by directly calling the register function of a specific
+ * implementation, e.g. console_16550_register() from <uart_16550.h>. Consoles
+ * registered that way can be unregistered/reconfigured with below functions.
+ */
+/* Remove a single console_t instance from the console list. */
+int console_unregister(console_t *console);
+/* Set scope mask of a console that determines in what states it is active. */
+void console_set_scope(console_t *console, unsigned int scope);
+
+/* Switch to a new global console state (CONSOLE_FLAG_BOOT/RUNTIME/CRASH). */
+void console_switch_state(unsigned int new_state);
+/* Output a character on all consoles registered for the current state. */
 int console_putc(int c);
+/* Read a character (blocking) from any console registered for current state. */
 int console_getc(void);
+/* Flush all consoles registered for the current state. */
 int console_flush(void);
 
+#if !MULTI_CONSOLE_API
+/* DEPRECATED on AArch64 -- use console_<driver>_register() instead! */
+int console_init(uintptr_t base_addr,
+		 unsigned int uart_clk, unsigned int baud_rate) __deprecated;
+void console_uninit(void) __deprecated;
```

Comment:
Build failures with the uniphier platform configs - e.g. build options DEBUG=1 PLAT=uniphier

/work/workspace/trusted-firmware-build/BUILD_CONFIG/uniphier-default/MODE/Debug/label/platforms/plat/socionext/uniphier/uniphier_console.S:211: undefined reference to `console_init'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/uniphier-default/MODE/Debug/label/platforms/plat/socionext/uniphier/uniphier_console.S:211:(.text.uniphier_console_setup+0x30): relocation truncated to fit: R_AARCH64_JUMP26 against undefined symbol `console_init'
make: *** [build/uniphier/debug/bl1/bl1.elf] Error 1

Adding MULTI_CONSOLE = 1 and also remove the __deprecated from console_uinit resolved the build issue.
(NOTE - the uniphier platform has ERROR_DEPRECATED=1)

Reply:
We need to remove the ERROR_DEPRECATED=1 from the upstream uniphier platform. This flag will always cause problems with deprecated APIs. We have a patch to do this in another in-flight patchset. Platforms should not be forced to use MULTI_CONSOLE=1.

Reply:
So, should I wait for that patch to land? Is it going to land soon, or if not can they pull it out into a separate PR?

Reply:
You can ignore that error in this PR. We'll deal with it separately

Reply:
@jwerner-chromium 
That PR has now landed and so this issue should be resolved.